### PR TITLE
FIX Correctly set duplicated has_one records for localised records

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

When duplicating a localised parent record, any `has_one` relations that should be localised with it still have the child ID of the original parent record. This means any changes to the duplicate child also apply to the original child because they're the same record.

The following configuration all needs to be set on the parent record, listing the child has_one relation in all of them
- `field_include`
- `localised_copy`
- `cascade_duplicates`

## Manual testing steps

1. Create a record with a `has_one` relation and localise that relation. I used elemental applied to a page, but the below config applied to the parent class should work:
    ```php
    private static array $has_one = [
        'MyRelation' => MyDataObject::class,
    ];

    private static array $field_include = [
        'MyRelationID',
    ];
    
    private static array $localised_copy = [
        'MyRelation',
    ];
    
    private static array $cascade_duplicates = [
        'MyRelation',
    ];

    private static array $extensions = [
        FluentExtension::class,
    ];
    ```
1. Create a locale in the CMS through `admin/locales`
1. Make sure you swap to the new locale in the CMS
1. Create the parent record in the CMS
1. Add a child record in the CMS
1. Duplicate the parent record.

Without this PR, the child attached to the new duplicate record is the _same_ as the one attached to the original, even though it should be a new one.

## Issues

- https://github.com/tractorcow-farm/silverstripe-fluent/issues/931

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/Extension/FluentExtension.php [hunk 1]
```diff
@@ -636,7 +637,18 @@ public function onAfterWrite(): void
      */
     public function onAfterDuplicate($original, $doWrite, $relations): void
     {
+        $owner = $this->getOwner();
         $localisedTables = $this->owner->getLocalisedTables();
+        // Get the names of all has_one columns that will have new IDs
+        $copyRelations = (array) $owner->config()->get('localised_copy');
```
Comment (mfendeksilverstripe) [inline src/Extension/FluentExtension.php:643]: Worth adding a check for `localisedCopyActive`?

File: src/Extension/FluentExtension.php [hunk 2]
```diff
@@ -654,6 +666,16 @@ public function onAfterDuplicate($original, $doWrite, $relations): void
...
                     WHERE \"RecordID\" = ?", [$toID, $fromID]);
+
+            // Make sure to update any has_one IDs - otherwise the new localised table entry will
+            // have the has_one ID from the original record.
+            $fieldsToUpdate = [];
+            foreach (array_intersect($fields, $copyHasOneRelations) as $copyFieldName) {
+                $fieldsToUpdate[$copyFieldName] = $owner->$copyFieldName;
+            }
+            if (!empty($fieldsToUpdate)) {
+                SQLUpdate::create($localisedTable, $fieldsToUpdate, ['RecordID' => $toID])->execute();
```
Comment (mfendeksilverstripe) [inline src/Extension/FluentExtension.php:679]: This is technically functional but I recommend something that is more ORM friendly - an approach similar to what we already have in `onBeforeWrite`. In our case we should be able to use `onBeforeDuplicate` extension point to achieve this.

**Review by mfendeksilverstripe — CHANGES_REQUESTED** [review]
Current approach is alright but I made a suggestion to try a more ORM friendly approach.

File: src/Extension/FluentExtension.php (hunk 1 above)
Comment (ChloeSartorelli) [reply src/Extension/FluentExtension.php:643]: I'm not sure what that is or how it works. Can you please check if this is also a problem when that is used, and if so, add the necessary context and replication steps to the issue?

File: src/Extension/FluentExtension.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/Extension/FluentExtension.php:679]: I'm using a low-level API intentionally to mirror the `insert` query above. Ideally setting these values would all be done in a single call, but I wasn't sure how to merge using the correct values into the existing `insert this row based on that existing row` query.

I'm not sure how `onBeforeDuplicate` applies here - can you please explain further?

**Comment by mfendeksilverstripe** [comment]
Looking into other approach (draft only so far):

https://github.com/tractorcow-farm/silverstripe-fluent/pull/933

File: src/Extension/FluentExtension.php (hunk 1 above)
Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:643]: Had a more detailed look on this one and I think it's fine to not have it as the intended case doesn't apply here.
 Context: this property allows you do avoid content duplication when models are being localised for some edge cases like in unit tests. This is not relevant for duplication scenario when a new model is being created.

File: src/Extension/FluentExtension.php (hunk 2 above)
Comment (mfendeksilverstripe) [reply src/Extension/FluentExtension.php:679]: Tried out a different approach using ORM as raw queries should be used only when there is no other option. It turns out that ORM approach has some other issues so I think your approach is overall better. Happy to keep this as is.

File: tests/php/Extension/LocalisedCopyTest/DuplicationTest.php [hunk 3]
```diff
@@ -238,6 +238,31 @@ public function testEditWithInheritedRelation(string $locale, bool $duplicated,
...
+            // Re-select both horses so we are getting what's in the DB not just what's in memory
+            $horse = Horse::get()->byID($horse->ID);
+            $duplicateHorse = Horse::get()->byID($duplicateHorse->ID);
+            $this->assertNotSame($horse, $duplicateHorse->TailID);
+            $this->assertNotSame($originalTailID, $duplicateHorse->TailID);
+            $this->assertNotSame($horse->TailID, $duplicateHorse->TailID);
+            $this->assertSame($originalTailID, $horse->TailID);
+        });
+    }
+
```
Comment (mfendeksilverstripe) [inline tests/php/Extension/LocalisedCopyTest/DuplicationTest.php:274]: Added a more detail test, specifically looking for data redundancy.

```php
/**
 * case: duplicate() called on localised record
 * desired outcome: has_one duplication is copied correctly to localised table
 */
public function testDuplicate(): void
{
    FluentState::singleton()->withState(function (FluentState $state): void {
        $state->setLocale('en_NZ');

        /** @var Horse|FluentExtension $originalHorse */
        $originalHorse = $this->objFromFixture(Horse::class, 'horse1');

        $tail = $originalHorse->Tail();
        $originalTailID = $tail->ID;

        $horseCountBefore = Horse::get()->count();
        $tailsCountBefore = Tail::get()->count();

        $duplicateHorse = $originalHorse->duplicate();

        // Re-fetch the original horse so we are certain we are asserting on the latest data
        $originalHorse = Horse::get()->byID($originalHorse->ID);

        $horseCountAfter = Horse::get()->count();
        $tailsCountAfter = Tail::get()->count();

        $this->assertEquals($horseCountBefore + 1, $horseCountAfter, 'We expect to see only one new horse after duplication');
        $this->assertEquals($tailsCountBefore + 1, $tailsCountAfter, 'We expect to see only one new tail after duplication');

        // Re-select both horses so we are getting what's in the DB not just what's in memory
        $originalHorse = Horse::get()->byID($originalHorse->ID);
        $duplicateHorse = Horse::get()->byID($duplicateHorse->ID);

        $this->assertNotSame($originalHorse->ID, $duplicateHorse->ID, 'We expect original and duplicated horse to have different IDs');
        $this->assertNotSame($originalHorse->TailID, $duplicateHorse->TailID, 'We expect original and duplicated tail to have different IDs');
        $this->assertSame($originalTailID, $originalHorse->TailID, 'We expect the original horse to still have the original tail');
    });
}
```

File: src/Extension/FluentExtension.php [hunk 4]
```diff
@@ -654,6 +666,16 @@ public function onAfterDuplicate($original, $doWrite, $relations): void
                     SELECT ? AS \"RecordID\", \"Locale\", $fields_str
                     FROM \"$localisedTable\"
                     WHERE \"RecordID\" = ?", [$toID, $fromID]);
+
+            // Make sure to update any has_one IDs - otherwise the new localised table entry will
+            // have the has_one ID from the original record.
+            $fieldsToUpdate = [];
+            foreach (array_intersect($fields, $copyHasOneRelations) as $copyFieldName) {
+                $fieldsToUpdate[$copyFieldName] = $owner->$copyFieldName;
```
Comment (mfendeksilverstripe) [inline src/Extension/FluentExtension.php]: Can you please add to this comment why the model (owner) in this has the correct data? I think that's the bit that's missing here.

Comment (mfendeksilverstripe) [inline src/Extension/FluentExtension.php]:
```suggestion
                $fieldsToUpdate[$copyFieldName] = $owner->{$copyFieldName};
```
[nitpick]

**Comment by mfendeksilverstripe** [comment]
Hey @GuySartorelli , I finished my testing and assessment. I think your approach is overall better so I closed https://github.com/tractorcow-farm/silverstripe-fluent/pull/933.

I left some minor comments to action but overall your solution works great. Tested on my local and I'm getting the desired outcome (new elemental area ID is assigned with correct contents).

Screenshot 2025-03-11 at 9 24 54 AM

**Comment by GuySartorelli** [comment]
Thanks for taking the time to review, Mojmir. I'll make those requested changes today.

File: tests/php/Extension/LocalisedCopyTest/DuplicationTest.php (hunk 3 above)
Comment (ChloeSartorelli) [reply tests/php/Extension/LocalisedCopyTest/DuplicationTest.php:274]: Done (copied in the new stuff and updated to match the style of the existing stuff)
Didn't re-fetch the original horse twice, though, as that doesn't seem necessary.

File: src/Extension/FluentExtension.php (hunk 4 above)
Comment (ChloeSartorelli) [reply src/Extension/FluentExtension.php]: Done

Comment (ChloeSartorelli) [reply src/Extension/FluentExtension.php]: Done

**Review by mfendeksilverstripe — APPROVED** [review]
Nice work 👍 GTG 🚀

**Comment by emteknetnz** [comment]
@mfendeksilverstripe You're probably allowed to merge things you approve :-) You have better knowledge of this PR and module in general than I do.
